plugins/generator-1.19.4/forge-1.19.4/templates/livingentity/livingentity.java.ftl

```diff
@@ -216,6 +232,109 @@ public class ${name}Entity extends ${extendsClass} <#if data.ranged>implements R
    	}
 	</#if>
 
+	<#if data.canTrade>
+	protected void updateTrades() {
+		<#assign professions = w.exclude(w.filterBrokenReferences(data.professionTrade), "WANDERING_TRADER")>
+		List<VillagerProfession> professions = List.of(
+			<#list professions as profession>
+				${profession}<#if profession?has_next>,</#if>
+			</#list>
+		);
+		Int2ObjectMap<VillagerTrades.ItemListing[]> trades = new Int2ObjectOpenHashMap<>();
+		VillagerTrades.TRADES.forEach((key, value) -> {
+			if (<#if data.excludeProfessions>!</#if>(professions.contains(key))) {
+				value.int2ObjectEntrySet().forEach(ent -> trades.put(ent.getIntKey(), Arrays.copyOf(ent.getValue(), ent.getValue().length)));
+			}
+		});
+		<#if data.professionTrade?contains("WanderingTrader") && !data.excludeProfessions>
+		VillagerTrades.WANDERING_TRADER_TRADES.int2ObjectEntrySet().forEach(e -> trades.put(e.getIntKey(), Arrays.copyOf(e.getValue(), e.getValue().length)));
+		</#if>
+		if (trades != null && !trades.isEmpty()) {
+			VillagerTrades.ItemListing[] leveledTrades = trades.get(<#if data.villagerTradingType>this.getVillagerData().getLevel()<#else>1</#if>);
+			if (leveledTrades != null) {
+				this.addOffersFromItemListings(this.getOffers(), leveledTrades, 2);
+			}
+		}
+	}
+
+	<#if data.restockCondition?? && (hasProcedure(data.restockCondition) || data.restockCondition.getFixedValue())>
+	<#if data.villagerTradingType>@Override</#if> public boolean canRestock() {
+		<#if hasProcedure(data.restockCondition)>
+			Entity entity = this;
+			Level world = entity.level;
+			double x = entity.getX();
+			double y = entity.getY();
+			double z = entity.getZ();
+		</#if>
+		return <@procedureOBJToConditionCode data.restockCondition true false/>;
+	}
+	</#if>
+
+	<#if data.villagerTradingType>
+	@Override protected void customServerAiStep() {

```

Comment:
As this is probably more or less copy+paste of vanilla code, I would prefer if Minecraft Code Provider was used for this

Reply:
Idk how to use Minecraft Code Provider...

Reply:
You can check other ME templates for examples, dimension uses it a lot in teleporter code